conformance_test.go

```diff
@@ -0,0 +1,157 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"path"
+	"path/filepath"
+	"testing"
+
+	"github.com/cucumber/godog"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+
+	"github.com/kubernetes-sigs/ingress-controller-conformance/test/conformance/defaultbackend"
+	"github.com/kubernetes-sigs/ingress-controller-conformance/test/files"
+	"github.com/kubernetes-sigs/ingress-controller-conformance/test/kubernetes"
+)
+
+var (
+	exitCode int
+)
+
+var (
+	godogFormat        string
+	godogTags          string
+	godogStopOnFailure bool
+	godogNoColors      bool
+	godogOutput        string
+
+	manifests string
+)
+
+func TestMain(m *testing.M) {
+	// register flags from klog (client-go verbose logging)
+	klog.InitFlags(nil)
+
+	flag.StringVar(&godogFormat, "format", "pretty", "Set godog format to use")
+	flag.StringVar(&godogTags, "tags", "", "Tags for conformance test")
+	flag.BoolVar(&godogStopOnFailure, "stop-on-failure ", false, "Stop when failure is found")
+	flag.BoolVar(&godogNoColors, "no-colors", false, "Disable colors in godog output")
+	flag.StringVar(&manifests, "manifests", "./manifests",
+		"Directory where manifests for test applications or scenerarios are located")
+	flag.StringVar(&godogOutput, "output-directory", ".", "Output directory for test reports")
+	flag.StringVar(&kubernetes.IngressClassValue, "ingress-class", "conformance",
+		"Sets the value of the annotation kubernetes.io/ingress.class in Ingress definitions")
+
+	flag.Parse()
+
+	manifestsPath, err := filepath.Abs(manifests)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if !files.IsDir(manifestsPath) {
+		log.Fatalf("The specified value in the flag --manifests-directory (%v) is not a directory", manifests)
+	}
+
+	//manifests.Path = manifestsPath
+
+	kubernetes.KubeClient, err = setupSuite()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if err := kubernetes.CleanupNamespaces(kubernetes.KubeClient); err != nil {
+		log.Fatalf("error deleting temporal namespaces: %v", err)
+	}
+
+	if code := m.Run(); code > exitCode {
+		exitCode = code
+	}
+
+	os.Exit(exitCode)
+}
+
+func setupSuite() (*clientset.Clientset, error) {
+	c, err := kubernetes.LoadClientset()
+	if err != nil {
+		return nil, fmt.Errorf("error loading client: %v", err)
+	}
+
+	dc := c.DiscoveryClient
+
+	serverVersion, serverErr := dc.ServerVersion()
+	if serverErr != nil {
+		return nil, fmt.Errorf("unexpected server error retrieving version: %v", serverErr)
+	}
+
+	if serverVersion != nil {
+		// TODO: check minimum k8s version?
+		log.Printf("kube-apiserver version: %s", serverVersion.GitVersion)
+	}
+
+	return c, nil
+}
+
+var (
+	features = map[string]func(*godog.Suite){
+		"features/default_backend.feature": defaultbackend.FeatureContext,
+	}
```

Comment:
The map of features here bothers me a bit since I feel it's error-prone...

1. It would be nice if these were detected and generated as well. Would it require a global `FeatureContext`?
2. This would, in turn, remove the `make codegen` notice to "Please add features..." in `codegen.go`
3. `make verify-codegen` does not bother to look in this map. We could push some implemented features and not skip their execution.

Reply:
> make verify-codegen does not bother to look in this map

That was the intention. The goal of verify-codegen is to check the mapping between the feature files and the generated go code, not the order of the execution of the tests.

Reply:
Does the order of execution matter? We should be able to turn on the `Random` option on godog.

I mostly feel this array could be generated, or built at runtime. Users should not have to worry about adding a reference to their FeatureContext here.

Reply:
> Does the order of execution matter?

I don't have a preference for that. I am just following what is being done in k8s conformance tests where there is no randomness.

Reply:
> The map of features here bothers me a bit since I feel it's error-prone...

I am checking what is needed to verify/update the `features` map content

Reply:
> The map of features here bothers me a bit since I feel it's error-prone...
> 
> 2. This would, in turn, remove the `make codegen` notice to "Please add features..." in `codegen.go`
> 3. `make verify-codegen` does not bother to look in this map. We could push some implemented features and not skip their execution.

Done.

Reply:
Awesome! I had a much smoother experience when adding a .feature file.